### PR TITLE
feat(plonky2): Partial implementation of '--show-plonky2'

### PR DESCRIPTION
**Pull request by nickysn**
Introduces a plonky2 "assembler writer" and implements display of a large number of plonky2 operations when the '--show-plonky2' command line option is used. Not all operations are shown, yet. The rest are going to be implemented in future commits.

### CODE REVIEW

File: compiler/noirc_evaluator/src/ssa/plonky2_gen/mod.rs [hunk 1]
```diff
@@ -294,24 +297,34 @@ impl P2Target {
     }
 }
 
-pub(crate) struct Builder {
-    builder: P2Builder,
+pub(crate) struct Builder<TAsmWriter: AsmWriter> {
```
Comment (smanilov) [inline compiler/noirc_evaluator/src/ssa/plonky2_gen/mod.rs:300]: Is it necessary for the Builder to be parametrized?

**Review by smanilov — COMMENTED** [review]
Looking good, just wondering about the type parameter.

Also, could we have a test, please? :)

File: compiler/noirc_evaluator/src/ssa/plonky2_gen/mod.rs (hunk 1 above)
Comment (nickysn) [reply compiler/noirc_evaluator/src/ssa/plonky2_gen/mod.rs:300]: If we use a trait with compile-time dispatch, yes. Alternatively, we can use a trait with dynamic dispatch, which uses a VMT table, but that is slightly slower and needs some nontrivial fighting with Rust's borrow checker in order to deal with the object's lifetimes. The third option is not to use a trait at all.

Comment (smanilov) [reply compiler/noirc_evaluator/src/ssa/plonky2_gen/mod.rs:300]: I was thinking about dynamic dispatch, I guess, but there's no need to fight the borrow checker, if it's resisting :)

**Comment by smanilov** [comment]
> > Also, could we have a test, please? :)
> 
> What kind of test do you have in mind? A simple test that just runs `nargo prove --show-plonky2` without checking the console output or something more complicated with output validation?
> 
> In principle, once the feature is completed, I'm thinking about writing a complex test that parses the "assembly" output and checks that there are no loose ends, e.g. reading from a target that wasn't initialized previously with a different instruction, etc. This will catch bugs, caused by missing instructions, but requires the --show-plonky2 output to be complete.
> 
> A simple test that doesn't check the output, just looks for successful termination is easy, but is it really valuable? We could also make a test that checks for a fixed expected output and then update it, as more instructions are printed with --show-plonky2. But it will also need to be changed anytime anything changes inside the compiler, for example in the RTL functions or in the early optimization passes.

I was thinking of a simple test, that shows what the printer outputs at the moment, so that I don't have to run it locally to find out. That would be most like the test that needs to be changed often, from the options you mentioned. If it's too much of a bother to keep it up to date we can delete it again, but it would really make reviewing easier, since I don't have to actually download and run the code. Or maybe I should not be so lazy :)

**Review by smanilov — APPROVED** [review]

**Comment by nickysn** [comment]
I'm attaching the output for zk_dungeon, so that you can see how the output looks like in this moment:
[zk_dungeon.plonky2.output.txt](https://github.com/user-attachments/files/15974184/zk_dungeon.plonky2.output.txt)

**Comment by smanilov** [comment]
> I'm attaching the output for zk_dungeon, so that you can see how the output looks like in this moment: [zk_dungeon.plonky2.output.txt](https://github.com/user-attachments/files/15974184/zk_dungeon.plonky2.output.txt)

Thanks, this is helpful, if a little long.

**Pull request merged**